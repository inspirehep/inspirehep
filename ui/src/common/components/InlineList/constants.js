export const DEFAULT_SEPARATE_ITEMS_CLASS = 'separate-items-with-comma';
export const SEPARATE_ITEMS_CLASSNAMES = [
  'separate-items-with-semicolon',
  'separate-items-with-comma',
  'separate-items-with-and',
  'separate-items-with-middledot',
];
