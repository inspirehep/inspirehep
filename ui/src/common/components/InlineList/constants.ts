export const SEPARATOR_SEMICOLON = '; ';
export const SEPARATOR_COMMA = ', ';
export const SEPARATOR_AND = ' and ';
export const SEPARATOR_MIDDLEDOT = ' \u2022 ';

export const SEPARATOR_TYPES = [
  SEPARATOR_SEMICOLON,
  SEPARATOR_COMMA,
  SEPARATOR_AND,
  SEPARATOR_MIDDLEDOT,
];

export const DEFAULT_SEPARATOR_TYPE = SEPARATOR_COMMA;
