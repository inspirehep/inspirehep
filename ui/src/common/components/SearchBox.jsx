import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Input } from 'antd';

import SearchScopeSelectContainer from '../containers/SearchScopeSelectContainer';

class SearchBox extends Component {
  static getDerivedStateFromProps(nextProps, prevState) {
    const { value } = nextProps;
    const { previousValue } = prevState;
    if (value !== previousValue) {
      return {
        ...prevState,
        previousValue: value,
        draftValue: value,
      };
    }
    return prevState;
  }

  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
    this.onSearch = this.onSearch.bind(this);
    this.state = {};
  }

  onChange(event) {
    const { value } = event.target;
    this.setState({ draftValue: value });
  }

  onSearch(value) {
    const { onSearch, namespace } = this.props;
    onSearch(namespace, value);
  }

  render() {
    const { placeholder } = this.props;
    const { draftValue } = this.state;
    return (
      <Input.Search
        autoFocus
        style={{ verticalAlign: 'middle' }}
        addonBefore={<SearchScopeSelectContainer />}
        placeholder={placeholder}
        value={draftValue}
        onChange={this.onChange}
        size="large"
        onSearch={this.onSearch}
        enterButton
      />
    );
  }
}

SearchBox.propTypes = {
  namespace: PropTypes.string.isRequired,
  onSearch: PropTypes.func.isRequired,
  value: PropTypes.string,
  placeholder: PropTypes.string,
};

SearchBox.defaultProps = {
  value: null,
  placeholder: null,
};

export default SearchBox;
