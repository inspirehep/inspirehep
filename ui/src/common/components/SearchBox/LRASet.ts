export default class LRASet {
  constructor(values, capacity) {
    this.capacity = capacity;
    this.values = [];
    this.valueToIndex = new Map();

    if (values) {
      values.forEach(this.add.bind(this));
    }
  }

  filter(filterFn) {
    return this.values.filter(filterFn);
  }

  add(value) {
    if (this.valueToIndex.has(value)) {
      this.remove(value);
    } else if (this.values.length >= this.capacity) {
      this.removeFirst();
    }

    this.values.push(value);
    this.valueToIndex.set(value, this.values.length - 1);
  }

  remove(value) {
    const currentIndex = this.valueToIndex.get(value);
    this.values.splice(currentIndex, 1);
    this.valueToIndex.delete(value);
  }

  removeFirst() {
    const removed = this.values.shift();
    this.valueToIndex.delete(removed);
  }

  toJSON() {
    return this.values;
  }
}
