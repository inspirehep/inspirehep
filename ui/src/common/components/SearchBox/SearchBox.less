.__SearchBox__ {
  .autocomplete {
    width: 100%;
    // HACK: doesn't set className from `Input.Search` to inner `input` for some reason
    input {
      border-top-left-radius: 0;
      border-bottom-left-radius: 0;
    }
  }
}
