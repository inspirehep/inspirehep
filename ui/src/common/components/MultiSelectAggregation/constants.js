import { RANK_VALUE_TO_DISPLAY } from '../../constants';

// moved separately from the component file to make it easy to mock in tests
export const SELECT_VALUE_TO_DISPLAY_MAPS_FOREACH_AGG = {
  Rank: RANK_VALUE_TO_DISPLAY,
};
