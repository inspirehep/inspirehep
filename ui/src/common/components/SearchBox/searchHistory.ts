import { SEARCH_BOX_NAMESPACES } from '../../../search/constants';
import LRASet from './LRASet';
import storage from '../../storage';

const STORAGE_KEY = 'search-box-history';
const HISTORY_LIMIT = 25;

function runWhenIdle(task) {
  if ('requestIdleCallback' in window) {
    window.requestIdleCallback(
      () => {
        task();
      },
      { timeout: 10000 }
    );
  } else {
    // TODO: find a good polyfill instead like: https://github.com/facebook/react/pull/8833
    task();
  }
}

export function persistHistory(history) {
  runWhenIdle(async () => {
    await storage.set(STORAGE_KEY, history);
  });
}

export function readHistory(callback) {
  return runWhenIdle(async () => {
    const historyFromStorage = (await storage.get(STORAGE_KEY)) || {};

    const history = {};
    SEARCH_BOX_NAMESPACES.forEach(namespace => {
      history[namespace] = new LRASet(
        historyFromStorage[namespace],
        HISTORY_LIMIT
      );
    });
    callback(history);
  });
}
