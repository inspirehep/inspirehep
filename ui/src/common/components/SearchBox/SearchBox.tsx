import React, { useCallback, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Input, AutoComplete } from 'antd';
import { HistoryOutlined } from '@ant-design/icons';
import classNames from 'classnames';
import './SearchBox.scss';
import SearchScopeSelectContainer from '../../containers/SearchScopeSelectContainer';
import { readHistory, persistHistory } from './searchHistory';
import IconText from '../IconText';

let HISTORY_BY_NAMESPACE = {};
readHistory(history => {
  HISTORY_BY_NAMESPACE = history;
});

function SearchBox({ value, placeholder, onSearch, namespace, className }) {
  const [inputValue, setInputValue] = useState(value);
  const [autoCompleteOptions, setAutoCompleteOptions] = useState([]);
  const [shouldSearch, setShouldSearch] = useState(false);

  // always override inputValue when `value` prop is changed
  useEffect(
    () => {
      // set empty string if `value` is null to force clearing text value of internal Input field.
      const valueOrEmpty = value != null ? value : '';
      setInputValue(valueOrEmpty);
    },
    [value]
  );
  const onInputChange = useCallback(
    event => {
      const newValue = event.target.value;
      setInputValue(newValue);
      const namespaceHistory = HISTORY_BY_NAMESPACE[namespace];
      if (newValue && HISTORY_BY_NAMESPACE[namespace]) {
        const options = namespaceHistory
          .filter(searchQuery => searchQuery.includes(newValue))
          .map(searchQuery => ({
            value: searchQuery,
            label: <IconText icon={<HistoryOutlined />} text={searchQuery} />,
          }));
        setAutoCompleteOptions(options);
      } else {
        setAutoCompleteOptions([]);
      }
    },
    [namespace]
  );
  const onInputSearch = useCallback(() => {
    setShouldSearch(true);
  }, []);

  const onAutoCompleteSelect = useCallback(selectedValue => {
    setInputValue(selectedValue);
    setShouldSearch(true);
  }, []);

  useEffect(
    () => {
      // FIXME: do it without an effect.
      // HACK: to perform a search only once after `Input.Search.onSearch` or `AutoComplete.onSelect`
      // since `onSelect` is called after `onSearch`, we can't do this `onSearch`.
      if (shouldSearch) {
        onSearch(namespace, inputValue);

        if (inputValue) {
          HISTORY_BY_NAMESPACE[namespace].add(inputValue.trim());
          persistHistory(HISTORY_BY_NAMESPACE);
        }
        setShouldSearch(false);
        setAutoCompleteOptions([]);
      }
    },
    [shouldSearch] // eslint-disable-line react-hooks/exhaustive-deps
  );

  return (
    // HACK: autocomplete select and scope select interferes with each other
    // when search scope is rendered using `Input.addonBefore` prop, inside autocomplete
    // that's why it's pulled out, and classNames are set manually to make it look like an addon
    <span className={classNames('__SearchBox__ ant-input-group', className)}>
      <span className="ant-input-group-addon">
        <SearchScopeSelectContainer />
      </span>
      <AutoComplete
        autoFocus
        dropdownClassName="header-dropdown"
        className="autocomplete"
        value={inputValue}
        dropdownMatchSelectWidth
        options={autoCompleteOptions}
        onSelect={onAutoCompleteSelect}
      >
        <Input.Search
          data-test-id="search-box-input"
          placeholder={placeholder}
          size="large"
          enterButton
          onChange={onInputChange}
          onSearch={onInputSearch}
        />
      </AutoComplete>
    </span>
  );
}

SearchBox.propTypes = {
  namespace: PropTypes.string.isRequired,
  className: PropTypes.string,
  onSearch: PropTypes.func.isRequired,
  value: PropTypes.string,
  placeholder: PropTypes.string,
};

export default SearchBox;
