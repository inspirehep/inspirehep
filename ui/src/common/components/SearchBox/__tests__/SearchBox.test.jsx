import { fireEvent } from '@testing-library/react';

import SearchBox from '../SearchBox';
import { LITERATURE_NS } from '../../../../search/constants';
import { renderWithProviders } from '../../../../fixtures/render';

describe('SearchBox', () => {
  it('render initial state with all props set', () => {
    const { asFragment } = renderWithProviders(
      <SearchBox
        namespace={LITERATURE_NS}
        value="value"
        placeholder="placeholder"
        searchScopeName="scope"
        onSearch={jest.fn()}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders new value on change', () => {
    const { getByTestId } = renderWithProviders(
      <SearchBox value="value" namespace={LITERATURE_NS} onSearch={jest.fn()} />
    );
    const input = getByTestId('search-box-input');
    fireEvent.change(input, { target: { value: 'new' } });
    expect(input).toHaveValue('new');
  });

  it('overrides internal state with prop', () => {
    const { rerender, getByTestId } = renderWithProviders(
      <SearchBox
        value="internal"
        namespace={LITERATURE_NS}
        onSearch={jest.fn()}
      />
    );

    const input = getByTestId('search-box-input');

    fireEvent.change(input, { target: { value: 'internal' } });
    expect(input).toHaveValue('internal');

    rerender(
      <SearchBox value="prop" namespace={LITERATURE_NS} onSearch={jest.fn()} />
    );

    expect(input).toHaveValue('prop');
  });
});
