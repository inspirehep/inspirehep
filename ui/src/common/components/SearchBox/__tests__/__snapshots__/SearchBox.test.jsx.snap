// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox overrides internal state with prop 1`] = `
<Fragment>
  <span
    className="__SearchBox__ ant-input-group"
  >
    <span
      className="ant-input-group-addon"
    >
      <Connect(SearchBoxNamespaceSelect) />
    </span>
    <ForwardRef(AutoComplete)
      autoFocus={true}
      className="autocomplete"
      dropdownClassName="header-dropdown"
      dropdownMatchSelectWidth={true}
      onSelect={[Function]}
      options={Array []}
      value="internal"
    >
      <Search
        data-test-id="search-box-input"
        enterButton={true}
        onChange={[Function]}
        onSearch={[Function]}
        size="large"
      />
    </ForwardRef(AutoComplete)>
  </span>
</Fragment>
`;

exports[`SearchBox render initial state with all props set 1`] = `
<Fragment>
  <span
    className="__SearchBox__ ant-input-group"
  >
    <span
      className="ant-input-group-addon"
    >
      <Connect(SearchBoxNamespaceSelect) />
    </span>
    <ForwardRef(AutoComplete)
      autoFocus={true}
      className="autocomplete"
      dropdownClassName="header-dropdown"
      dropdownMatchSelectWidth={true}
      onSelect={[Function]}
      options={Array []}
      value="value"
    >
      <Search
        data-test-id="search-box-input"
        enterButton={true}
        onChange={[Function]}
        onSearch={[Function]}
        placeholder="placeholder"
        size="large"
      />
    </ForwardRef(AutoComplete)>
  </span>
</Fragment>
`;

exports[`SearchBox renders new value on change 1`] = `
<Fragment>
  <span
    className="__SearchBox__ ant-input-group"
  >
    <span
      className="ant-input-group-addon"
    >
      <Connect(SearchBoxNamespaceSelect) />
    </span>
    <ForwardRef(AutoComplete)
      autoFocus={true}
      className="autocomplete"
      dropdownClassName="header-dropdown"
      dropdownMatchSelectWidth={true}
      onSelect={[Function]}
      options={Array []}
      value="new"
    >
      <Search
        data-test-id="search-box-input"
        enterButton={true}
        onChange={[Function]}
        onSearch={[Function]}
        size="large"
      />
    </ForwardRef(AutoComplete)>
  </span>
</Fragment>
`;
